[FLINK-15981][network] Implement FileRegion way to shuffle file-based blocking partition in network stack

zhijiangW opened the pull request:
## What is the purpose of the change

Implement the `FileRegion` based way in netty stack for transferring the bounded blocking partition data with file type, then we can avoid introducing any memory overhead in application layer to resolve OOM issue before.

Besides that, compared with existing mmap option, `FileRegion` has some advantages as below:

- Reduce the context switch

- Reduce the memory copy from kernel to socket based on DMA support 

- Somehow avoid killing by yarn cluster, since mmap has already side effect to be killed by yarn because of memory overuse.(`FileRegion` option needs to be further experimented in real scenario for this concern).

## Brief change log

- Refactor the `NettyMessage` as interface instead of abstract class.

- Introduce `FileRegionResponse` to implement `NettyMessage` for making use of file region way provided by netty stack.

- Introduce `PartitionData` and `BoundedPartitionData` abstract classes for unifying the code paths between pipelined and blocking partitions.

- Refactor series of interface methods from `getBuffer` to `getData` for fitting the above changes.

- Request one additional memory segment from global pool in `SingleInputGate` for reading bounded partition data in local channel.

## Verifying this change

The previous unit tests should cover almost all the changes, might supplement some new unit tests if necessary future.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionData.java:
```diff
@@ -45,6 +51,14 @@
+	 * @param segment it might be needed to read the partition data into.
+	 * @return the buffer represents the partition data.
+	 */
+	public abstract Buffer getBuffer(@Nullable MemorySegment segment) throws IOException;
```
TBH, I am a bit torn and not satisfied with introducing this method, since the argument `segment` seem not very elegant from the caller usage. But I did not think of a better option for resolving it now. :(

StephanEwen replied:
I would change this to `Supplier<MemorySegment>` so that the code lazily fetches (and allocates) the memory segment only when needed. See also next comment about lazy creation.

zhijiangW commented:
FYI, the last four commits separation is only for easy review, will be squashed finally before merging.

zhijiangW commented:
@flinkbot run azure

zhijiangW commented:
@flinkbot run azure

zhijiangW commented:
I supplemented a new micro-benchmark case for verifying the change via [PR](https://github.com/apache/flink-benchmarks/pull/4), and the results as below seem no obvious change.

Before change:                                                                                                    Mode  Cnt     Score     Error   Units
RemoteBlockingPartitionThroughputBenchmark.remoteBlockingPartition  thrpt   30  7009.393 ± 251.786  ops/ms

After change:                                                                                                       Mode  Cnt     Score     Error   Units
RemoteBlockingPartitionThroughputBenchmark.remoteBlockingPartition  thrpt   30  7109.649 ± 197.832  ops/ms

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -238,6 +248,12 @@ public void setup() throws IOException {
 		setupChannels();
 
+		//TODO we can further judge the condition whether the type is file-based or mmap-based
+		if (consumedPartitionType.isBlocking()) {
```
I would simply make this a simple lazy-created MemorySegement, without any logic trying to determine what the shuffle type is. If this is fetched only if needed (see using `Supplier<MemorySegment>` above) then that will be good enough.

I think it would also be fine if this was an unpooled segment, not obtained from the `memorySegmentProvider` but just created on first access. Would make things simpler, you don't have to worry about recycling, races between creation and early disposal, etc.

zhijiangW replied:
Agree, it seems better to make it lazy-created and unpooled segment.

zhijiangW replied:
I thought of another issue to use `Supplier<MemorySegment>` way, that means for every call of `LocalInputChannel#getNextBuffer` we need to allocate a new segment for it (no matter with pooled or unpooled segment). My previous way tried to allocate only one segment for always reuse in all the ``LocalInputChannel#getNextBuffer` calls.

But I am fine with the proposal of using unpooled segment here, because it indeed simplify some aspects as you mentioned above.

StephanEwen reviewed (commented):
All in all the design looks good to me.

The are some small comments inline, mainly about some simplification of the code for `LocalInputChannel`.

Below are some of the more general issues we should double-check.

### Compatibility with SSL

I saw some code in Netty that looked like `FileRegion` might not be compatible with SSL. Would be good to verify that as a first step, because if it does indeed not work with SSL, we'd better know this soon.
Then we might have to evaluate whether reading into a buffer allocated from Netty can work, or whether we need some completely different approach here.

### Naming

This PR is renaming a lot of classes, to make the abstract base classes, for example `BufferAndAvailability` to `DataAndAvailability`.

In my opinion, we can keep the current name for the base class in pretty much all cases, because they are maybe not 100% correct, but still pretty understandable (thinking of a file region as a lazy buffer). That would lead to less renaming (and new names to understand for other committers working on that code).
We could then call the subclasses `ByteBufferAndAvailability` and `FileRegionAndAvailability`.

For the class `ResponseInfo` the name `ResponseMetadata`, `ResponseHeaderData` or even `BufferResponseHeader` might be more descriptive.

### An slightly different structure for Netty Messages

It looks to me like a lot of the complexity comes from the fact that the `NettyMessage` needs always to build one final object (`ByteBuf` or `FileRegion`) to be then written.

We can simplify things if we change the method
```java
abstract ByteBuf write(ByteBufAllocator allocator) throws Exception;
```
to
```java
abstract void write(ChannelOutboundInvoker out, ByteBufAllocator allocator) throws Exception;
```
Then the actual messages could just write one or more objects directly. The `BufferResponse` message would look like this (no need any more to use a CompositeBuffer):
```java
@Override
void write(ChannelOutboundInvoker out, ByteBufAllocator allocator) throws IOException {
	ByteBuf headerBuf = null;
	try {
		headerBuf = allocateBuffer(allocator, ID, MESSAGE_HEADER_LENGTH, bufferSize, false);
		receiverId.writeTo(headerBuf);
		headerBuf.writeInt(sequenceNumber);
		headerBuf.writeInt(backlog);
		headerBuf.writeByte(dataType.ordinal());
		headerBuf.writeBoolean(isCompressed);
		headerBuf.writeInt(buffer.readableBytes());

		out.write(headerBuf);

		// in order to forward the buffer to netty, it needs an allocator set
		buffer.setAllocator(allocator);
		out.write(buffer);
	}
	catch (Throwable t) {
		if (headerBuf != null) {
			headerBuf.release();
		}
		buffer.recycleBuffer();
		ExceptionUtils.rethrowIOException(t);
	}
}
```

We would not even need a separate subclass of `BufferResponse` any more, if we let the `buffer` field be an Object (either `ByteBuf` or `FileRegion`).

Then you also don't need a class that implements `FileRegion` any more, but you can simply use Netty's `DefaultFileRegion`.

What do you think about this change? It seems at a first glance like it would simplify stuff quite a bit. Or am I overlooking some detail here?

### Interface Changes

One other advantage that the change above would have is that you don't need to change `NettyMessage` to an interface (could remain an abstract class). In most cases it makes little difference, but if in doubt, abstract classes have cheaper method invocations. I think especially default methods are actually really expensive, unless the JIT can fully de-virtualize them.

zhijiangW commented:
Thanks for your pretty suggestive reviews @StephanEwen and I really get some profound, original insights from you!

>  Compatibility with SSL

I really overlooked the SSL branch before. After double verify, the file region way is indeed not compatible with SSL as you concerned, also mentioned in netty [issue](https://github.com/netty/netty/issues/2075). ATM I do not think of a proper way to support SSL with file region. Then my rough thought is to retain two branches to work around. The previous `FileBufferReader` implementation with memory overhead is still valid if SSL enabled. And I introduce another implementation of `BoundedData.Reader` for taking file region way without SSL option. 

I guess SSL not widely used in practice, so most of the scenarios can still benefit from this improvement of memory overhead. WDYT or have other better options?

> Naming

I ever a bit torn whether to retain the previous naming for the reason you mentioned above. I totally agree with your saying `thinking of a file region as a lazy buffer`, then +1 to keep the previous naming.

> An slightly different structure for Netty Messages

After double checking, I had the misunderstanding of the netty stack call before to result in current implementation. I wrongly thought that `NettyMessage#write` can only write `FileRegion` instance without any `ByteBuf` instance in order to guarantee the followup `FileRegion#transferTo` called properly.  So I only allocated header buffer in `NettyMessage#write` and write it together with `FileRegion#transferTo`, that is the previous consideration without reusing `DefaultFileRegion` directly.

Your above inputs of refactoring `NettyMessage#write` give me some insights to try out again and dispel my above misunderstanding. We can indeed write header ByteBuf to `ChannelOutboundInvoker` directly for both `BufferResponse` and `FileRegion` ways, then the code path is almost the same for them now. I conclude the following changes:

- Introduce another `FileRegionResponse` which extends `DefaultFileRegion`, because we need to override the `#deallocate` method not to close file channel for every call.

- Refactor current `NettyMessage#write` method into `write(ChannelOutboundInvoker out, ByteBufAllocator allocator, ChannelPromise promise)`, then return `void` for it. I find it will also bring some logics refactoring in some unit tests.

- Make `BufferResponse#buffer` as an object which can be either `Buffer` or `FileRegionResponse`, but we also need to consider how to refactor current usages like `buffer#recycleBuffer`, `buffer#setAllocator` and `BufferResponse#getBuffer` after `buffer` becoming an object. 

- Keep `NettyMessage` still as abstract class since we do not need to introduce extra `NettyMessage` subclass to reuse existing `BufferResponse` as above.

StephanEwen commented:
Thanks, @zhijiangW - some thoughts on your suggestions here:

> Introduce another FileRegionResponse which extends DefaultFileRegion, because we need to override the #deallocate method not to close file channel for every call.

Do we need to `FileRegionResponse` to extend `DefaultFileRegion`, or can the `FileRegionResponse` contain a `DefaultFileRegion` object? Given the complex interface of `FileRegion` in Netty (and how it may change between versions) might be good to avoid inheriting from it.

You are right about `buffer#recycleBuffer`, `buffer#setAllocator` and `BufferResponse#getBuffer` - we need to handle this differently for buffers and regions.

> Handling SSL

I think SSL is quite important, many setups need it for data security compliance reasons. So we definitely need to support it.
For the time being it is probably okay to say that the optimization from this PR here is not active when SSL is active, that we keep using the original version with the higher memory use when SSL is active. Then we see how the sort-based shuffle develops, maybe we need to encourage people to use that one when SSL is active.

zhijiangW commented:
Thanks for the further confirmation @StephanEwen !

> Do we need to FileRegionResponse to extend DefaultFileRegion, or can the FileRegionResponse contain a DefaultFileRegion object? Given the complex interface of FileRegion in Netty (and how it may change between versions) might be good to avoid inheriting from it.

As we confirmed before, we only need one `BufferResponse` message for both `Buffer` and `FileRegion` ways. I tried to introduce `FileRegionBuffer` to implement `Buffer` interface and it has the `FileRegion` inside. Then we do not need to touch any previous `Buffer` instances and compatible with relevant existing interfaces.

But we can not reuse `DefaultFileRegion` provided by netty directly, since the `DefaultFileRegion#deallocate()` would close underlying `FileChannel` for every transfer. So we have to extend `DefaultFileRegion` to override `#deallocate` method with empty operation. 

> You are right about buffer#recycleBuffer, buffer#setAllocator and BufferResponse#getBuffer - we need to handle this differently for buffers and regions.

I guess we do not need consider these issues now, as I assume `FileRegionBuffer` implement `Buffer` interface compatible with existing methods. Also we do not need any `Object` field in `BufferResponse` component.

> Handling SSL

Totally agree with your points for it.

FYI: My new version codes are almost ready to resolve all the above comments. I will update this PR probably early tomorrow after polishing the code a bit.

zhijiangW commented:
@StephanEwen, I have updated the codes for addressing above comments. 

Regards the tests, I think the existing `BoundedBlockingSubpartitionTest` and `BoundedBlockingSubpartitionWriteReadTest` already covered the local read part. 
I will further consider how to cover the remote network read by existing tests after you approve the current implementation way.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/FileRegionBuffer.java:
```diff
@@ -0,0 +1,213 @@
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Wrapper for {@link org.apache.flink.shaded.netty4.io.netty.channel.FileRegion} instance.
```
Would be good to explain a bit more why this class is like it is, and how it works:

  - This class implements `Buffer` so we can move it can be used like another `Buffer`, although it behaves "read-only" style, so many methods throw UnsupportedOperationException.

  - This extends from Netty's `DefaultFileRegion` so we don't need to do any special handling. Netty will internally treat this differently than the `Buffer` that implements `ByteBuf`.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartition.java:
```diff
@@ -74,7 +74,10 @@
+	private final Set<ResultSubpartitionView> readers;
+
+	/** Flag whether SSL is enabled for network transfer with file type. */
+	private final boolean fileWithoutSSLEnabled;
```
How about renaming this to `useDirectFileTransfer` ?

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionView.java:
```diff
@@ -0,0 +1,199 @@
+ * The reader (read view) of a BoundedBlockingSubpartition based on
+ * {@link org.apache.flink.shaded.netty4.io.netty.channel.FileRegion}.
+ */
+public class BoundedBlockingSubpartitionView implements ResultSubpartitionView {
```
Can we rename this to `BoundedBlockingSubpartitionDirectTransferReader`? Then the naming scheme is similar to `BoundedBlockingSubpartitionReader`.

It is a bit confusing when some two different implementation on the same level use different naming patterns.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/FileRegionBuffer.java:
```diff
@@ -0,0 +1,213 @@
+	 */
+	@Override
+	public ByteBuffer getNioBufferReadable() {
+		return BufferReaderWriterUtil.tryReadByteBuffer(fileChannel, bufferSize);
```
This method here is a little concerning, because if used wrong it voids the whole implementation. It reads into a newly allocated buffer.

I assume this is mainly needed for events?

One alternative here could be that the reading of events from the file channel produces a regular `NetworkBuffer`. Then this method here would not be needed.

zhijiangW replied:
As I mentioned in javadoc for this method, it is only used for unit tests ATM, not for the core codes.
Actually in the first version I made this method for `UnsupportedOperationException`, but I found many exiting tests rely on this method like `BoundedBlockingSubpartitionWriteReadTest`. So I implemented this method for avoiding refactoring the logic of unit tests for less efforts. Do you think we should refactor the unit test to avoid calling this way instead?

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionView.java:
```diff
@@ -0,0 +1,199 @@
+		@Nullable
+		@Override
+		public Buffer nextBuffer() throws IOException {
+			// As the file region transfer via network channel will not modify the underlining file position,
```
Maybe we can optimize this method a bit.

If you inline the code from BufferReaderWriterUtil.readFromByteChannel() here, then you can avoid having the logic tracking `lastPosition` etc, and simply advance the position always here.

It should also help to avoid some calls to `fileChannel.position()`, which are not cheap.

zhijiangW replied:
The current way is indeed a bit obscure. Actually it is involved in two different usages ATM:

- For network file transfer, the underlining file position for data part is not changed while transferring.

- For local input channel reading, it will change the respective underlining file position via `FileRegionBuffer#readInto`.

So when the `nextBuffer` is called here, it does not know the caller is from local channel or network stack, then it can not always advance the position. Here I rely on the `lastPosition` for implicitly distinguishing the different usages.

My first version PR seems a bit better here to always advance the position in the constructor of extended `FileRegion` instance, since it would be only used for network stack. But after the refactoring in the current version, I did not find another better place to advance the position reasonably. Or I misunderstood your proposed way?

StephanEwen replied:
I think this is very fragile, using the `lastPosition` to distinguish the uses.

Can we say that this part here (creating regions from the file) always advances the position, and the actual reading part never advances the position? For that, the `FileRegionBuffer#readInto(...)` would also need to take the offset/length fields stored in the region to handle this.

StephanEwen replied:
If you want, I can try and create a suggestion for this change.

zhijiangW replied:
Yeah, I agree with the fragile point. Let me consider how to refactor it a bit.

The `FileRegionBuffer#readInto` relies on `FileChannel#read(ByteBuffer dst)` to read the data part, which will automatically updates the position. If I always advance the data part position here before actual reading, then for local channel reading, I need to roll back the respective position. It might be a bit better than current way since we do not need to maintain the `lastPosition` variable. My understanding is right?

The realistic way is to only advance the position for network direct transfer, and leave the normal advance for local input channel while `FileChannel#read(ByteBuffer dst)`. I will further think whether it can be done in some place.

StephanEwen reviewed (commented):
Thanks, this looks pretty good to me. This is very close to being done, from my side, only I have only minor cleanup suggestions.

The general structure and abstraction looks good.
The `FileRegionBuffer` extending `Buffer` is a slight hack, but in this case there are some good reasons for it (Netty's internal logic). Still, we should explain in the JavaDocs why this is implemented like it is, to explain it to future developers.

I have some suggestions for minor changes to make some places a bit nicer, but this should not be much.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -371,6 +368,22 @@ ByteBuf write(ByteBufAllocator allocator) throws IOException {
 
+		private ByteBuf fillHeader(ByteBufAllocator allocator) {
+			// in order to forward the buffer to netty, it needs an allocator set
+			buffer.setAllocator(allocator);
```
I ever had the same feeling before exactly as you said. The deduplication seems not worth here.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -691,23 +672,12 @@ static CloseRequest readFrom(@SuppressWarnings("unused") ByteBuf buffer) throws
-				throw new IOException(t);
-			}
+		void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator allocator) throws IOException {
+			Consumer<ByteBuf> consumer = (bb) -> {
```
I indeed overlooked the performance concern for the frequent message.  The `handleException` deduplication seems pretty good for me.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferReaderWriterUtil.java:
```diff
@@ -187,6 +205,17 @@ static ByteBuffer allocatedHeaderBuffer() {
+			tryReadByteBuffer(channel, bb);
+			bb.flip();
+		} catch (IOException ex) {
+			ExceptionUtils.rethrow(ex);
```
Agree, I really had the same concern before. Actually my previous way was throwing explicit `IOException` here. But I found it also need to adjust the interface method `Buffer#getNioBufferReadable` for throwing `IOException`, even it would further bring up many serious of changes for exiting usages.

It was also caused by [comment](https://github.com/apache/flink/pull/13523#discussion_r516235008). Then let's confirm them together.

StephanEwen commented:
Thanks! I can address these remaining points and merge this.

StephanEwen commented:
Rebased and manually merged in 4869d3593d5b7fcff36abbe71a3e1a46b7a64998
